forest/cli/src/cli/snapshot_cmd.rs

```diff
@@ -323,6 +353,131 @@ fn clean(config: &Config, snapshot_dir: &Option<PathBuf>, force: bool) -> anyhow
     Ok(())
 }
 
+async fn validate(
+    config: &Config,
+    recent_stateroots: &i64,
+    snapshot: &PathBuf,
+    force: bool,
+) -> anyhow::Result<()> {
+    let confirm = if !force {
+        Confirm::with_theme(&ColorfulTheme::default())
+            .with_prompt(format!(
+                "This will result in using approximately {} MB of data. Proceed?",
+                std::fs::metadata(snapshot)?.len() / (1024 * 1024)
+            ))
+            .default(false)
+            .interact()
+            .unwrap_or_default()
+    } else {
+        true
+    };
+
+    if confirm {
+        let tmp_db_path = TempDir::new()?;
+
+        let db_path = tmp_db_path.path().join(&config.chain.name);
+
+        let db = RocksDb::open(db_path, &config.rocks_db)?;
+
+        let chain_store = Arc::new(ChainStore::new(db).await);
+
+        let genesis = read_genesis_header(
+            config.client.genesis_file.as_ref(),
+            config.chain.genesis_bytes(),
+            &chain_store,
+        )
+        .await?;
+
+        let cids = {
+            let file = tokio::fs::File::open(&snapshot).await?;
+            let reader = FetchProgress::fetch_from_file(file).await?;
+            load_car(chain_store.blockstore(), reader.compat()).await?
+        };
+
+        let ts = chain_store.tipset_from_keys(&TipsetKeys::new(cids)).await?;
+
+        validate_links_and_genesis_traversal(
+            &chain_store,
+            ts,
+            chain_store.blockstore(),
+            *recent_stateroots,
+            &genesis,
+        )
+        .await?;
+    }
+
+    Ok(())
+}
+
+async fn validate_links_and_genesis_traversal<DB>(
+    chain_store: &ChainStore<DB>,
+    ts: Arc<Tipset>,
+    db: &DB,
+    recent_stateroots: ChainEpoch,
+    genesis_tipset: &Tipset,
+) -> anyhow::Result<()>
+where
+    DB: fvm_ipld_blockstore::Blockstore + Store + Send + Sync,
+{
+    let mut seen = std::collections::HashSet::<Cid>::new();
+    let upto = ts.epoch() - recent_stateroots;
+
+    let mut tsk = ts.parents().clone();
+    // Security: Recursive snapshots are difficult to create but not impossible. This
+    // limits the amount of recursion we do.
+    let mut prev_epoch = ts.epoch();
+
+    let total_size = recent_stateroots;
```

Comment:
We have to scan through every tipset in the snapshot, not just the `recent_stateroots`. Scanning the recent stateroots is expensive but the cost is fixed. Scanning all the way to the genesis block gets more expensive as the snapshot grows.

Either include all the epochs in the progress bar or use two progress bars (one for the recent state roots and one for the genesis scanning).

Reply:
Went with the former